### DISCUSSION TITLE
Interest in building zxing-cpp with disabled exception handling?

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#710] [Polls]
Sparked by #708 I looked briefly into the subject of disabling exception handling (`-fno-exceptions` in gcc and clang speak). I see 3 options regarding this subject:

 1. exceptions are part of the API/ABI and compilation with disabled exception handling of either library and client code is not officially supported. (status quo)
 2. there is a build option for `zxing-cpp` that allows to disable exception handling for client code with some some major caveats, like only available for the reader API. (relatively easy to implement)
 3. there is a build option/API redesign that completely removes exception handling from the code base. (not going to happen)
 
Now my question is basically whether option 2 is worth spending time on?

**Comment by duanqn** [comment]
I read about `-fno-exceptions` [here](https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_exceptions.html) and it seems like it is a GNU extension to the standard C++. If so, then it should not be a requirement for ZXing-cpp to support a non-standard language extension.

It is also specifically noted in the manual that 

> If you might have some code that throws, you shouldn't use -fno-exceptions. If you have some code that uses try or catch, you shouldn't use -fno-exceptions.


Also, I think it is possible for any downstream app to consume ZXing-cpp with an extra wrapper layer, like
```
// wrapper.cpp
// compile with -fexceptions
try{
  // wrap all ZXing-cpp calls in here
}
catch(...){
  // Do something or do nothing at all
}
```

and compile the rest of their source files with `-fno-exceptions`

**Reply by duanqn** [reply]
Maybe we need to move any function implementation involving exceptions to .cpp files, though. Otherwise they would get compile errors when they include the header.
